File: mysql_examples/recipes/sequences/seqdemo.py
#!/usr/bin/python
# seqdiag.py - test AUTO_INCREMENT operations

import MySQLdb
import Cookbook

def init_table (conn):
  try:
    cursor = conn.cursor ()
    cursor.execute ("DROP TABLE IF EXISTS seqdiag")
    cursor.execute ("""
                  CREATE TABLE seqdiag
                  (seq INT UNSIGNED NOT NULL AUTO_INCREMENT PRIMARY KEY)
                """)
    cursor.close ()
  except MySQLdb.Error, e:
    print "Cannot create seqdiag table"
    print e.args

def gen_seq_val (cursor):
  try:
    cursor.execute ("INSERT INTO seqdiag (seq) VALUES(NULL)")
  except MySQLdb.Error, e:
    print "Cannot create AUTO_INCREMENT value"
    print e.args

conn = Cookbook.connect ()

init_table (conn)
cursor = conn.cursor ()

gen_seq_val (cursor)
seq = conn.insert_id ()
print "seq:", seq
cursor.execute ("SET @x = LAST_INSERT_ID(99)")
seq = conn.insert_id ()
print "seq after SET via insert_id():", seq
cursor.execute ("SELECT LAST_INSERT_ID()")
row = cursor.fetchone ()
seq = row[0]
print "seq after SET via LAST_INSERT_ID():", seq

cursor.close ()
conn.close ()

conn1 = Cookbook.connect ()
conn2 = Cookbook.connect ()

# Issue two sequence-generating queries using separate connection
# objects to see whether they interfere with each other's
# insert_id() value.
#@ _INDEPENDENT_CONNS_
cursor1 = conn1.cursor ()
cursor2 = conn2.cursor ()
gen_seq_val (cursor1)   # issue query that generates a sequence number
gen_seq_val (cursor2)   # issue another, using a different cursor
seq1 = conn1.insert_id ()
seq2 = conn2.insert_id ()
print "seq1:", seq1, "seq2:", seq2  # these values will be different
cursor1.close ()
cursor2.close ()
#@ _INDEPENDENT_CONNS_

conn1.close ()
conn2.close ()

File: mysql_examples/recipes/sequences/SeqDemo.java
// SeqDemo.java - demonstrate AUTO_INCREMENT operations

import java.sql.*;
import com.kitebird.mcb.Cookbook;

public class SeqDemo
{
  public static void initTable (Connection conn, String tblType)
  {
    try
    {
      Statement s = conn.createStatement ();
      s.execute ("DROP TABLE IF EXISTS seqdiag");
      s.execute (
            "CREATE TABLE seqdiag"
            + " (seq INT UNSIGNED NOT NULL AUTO_INCREMENT PRIMARY KEY)"
            + " ENGINE=" + tblType);
      s.close ();
// use SHOW CREATE TABLE here to see if type is correct
    }
    catch (Exception e)
    {
      Cookbook.printErrorMessage (e);
    }
  }

  public static void main (String[] args)
  {
    Connection conn = null;
    String tblType = "MyISAM";

    if (args.length > 0)
      tblType = args[0];
    System.out.println ("Testing using table type: " + tblType);

    try
    {
      conn = Cookbook.connect ();
      initTable (conn, tblType);

      // Check whether different Statement objects hold
      // independent insert-id values 

      String stmt = "INSERT INTO seqdiag (seq) VALUES(NULL)";
      Statement s = conn.createStatement ();
      Statement s2 = conn.createStatement ();
      s.executeUpdate (stmt);
      s2.executeUpdate (stmt);
      long seq = ((com.mysql.jdbc.Statement) s).getLastInsertID ();
      long seq2 = ((com.mysql.jdbc.Statement) s2).getLastInsertID ();
      System.out.println ("seq1: " + seq + ", seq2: " + seq2);
      if (seq == seq2)
        System.out.println ("VALUES SHOULD HAVE BEEN DIFFERENT!");
      s.close ();
      s2.close ();
    }
    catch (Exception e)
    {
      Cookbook.printErrorMessage (e);
    }
    finally
    {
      if (conn != null)
      {
        try
        {
          conn.close ();
        }
        catch (Exception e) { }
      }
    }
  }
}
